src/concurrency/data_race.rs

```diff
@@ -1044,6 +982,105 @@ trait EvalContextPrivExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
         result
     }
 
+    /// Checks that an atomic access is legal at the given place.
+    fn atomic_access_check(&self, place: &MPlaceTy<'tcx, Provenance>) -> InterpResult<'tcx> {
+        let this = self.eval_context_ref();
+        // Check alignment requirements. Atomics must always be aligned to their size,
+        // even if the type they wrap would be less aligned (e.g. AtomicU64 on 32bit must
+        // be 8-aligned).
+        let align = Align::from_bytes(place.layout.size.bytes()).unwrap();
+        this.check_ptr_access_align(
+            place.ptr,
+            place.layout.size,
+            align,
+            CheckInAllocMsg::MemoryAccessTest,
+        )?;
+        // Ensure the allocation is mutable. Even failing (read-only) compare_exchange need mutable
+        // memory on many targets (i.e., they segfault if taht memory is mapped read-only), and
+        // atomic loads can be implemented via compare_exchange on some targets. See
+        // <https://github.com/rust-lang/miri/issues/2463>.
+        // We avoid `get_ptr_alloc` since we do *not* want to run the access hooks -- the actual
+        // access will happen later.
+        let (alloc_id, _offset, _prov) =
+            this.ptr_try_get_alloc_id(place.ptr).expect("there are no zero-sized atomic accesses");
+        if this.get_alloc_mutability(alloc_id)? == Mutability::Not {
+            throw_ub_format!("atomic operations cannot be performed on read-only memory");
```

Comment:
Is it worth noting that this is target/size-specific for atomic loads? Or maybe there's a way to phrase it less strictly?

The guidance I've always been given for dealing with cross-process shared memory is that this is that atomics *must* be used if another process may write to it and that raw pointers (since obviously references are out of the question) would be UB (even if volatile were used), as it would still be a data race.

I do sympathize with not writing to add the target-specific logic for when this is broken to miri (it seems fragile and gross), but my concern is that encouraging people to *not* use atomics for this would lead them to perform worse UB.

(Unless we want to consider it not a data race?)

Reply:
AFAIK, the following are cases where atomic loads perform operations that could be problematic on read-only memory. [^others]

- aarch64: 128-bit loads ([LL/SC by ld(a)xp/st(l)xp](https://godbolt.org/z/1rjfvsaME)) (if FEAT_LSE2 which makes load pair/store pair single-copy atomic is not available)
- x86_64: 128-bit loads ([CAS by cmpxchg16b](https://godbolt.org/z/xnM4Y1MP8))
- x86: 64-bit loads ([CAS by cmpxchg8b](https://godbolt.org/z/37MjEnz5z)) (if both SSE and X87 are not available)
- arm (pre-v6, Linux only): any non-relaxed/non-unordered loads ([CAS by `__kuser_cmpxchg` via `__sync_*` builtins](https://godbolt.org/z/6GEYsnnPP))

Considering the above, I think it is ok to allow atomic load on read-only memory when all of the following are met. (assuming we don't want to do complex and fragile target-specific processing) 

- size is pointer-width or smaller. [^ptr]
- ordering is relaxed (or unordered).

And the error message would probably be something like:

```
atomic load operations with non-relaxed ordering or grater than pointer-width cannot be performed on read-only memory
```

(Even if Miri does not allow them, additional notes would be helpful.)


[^others]: [armv6+'s 64-bit load is `ldrexd`](https://godbolt.org/z/oGKjdErW3), [powerpc64 and s390x have 128-bit atomic load instructions](https://github.com/taiki-e/portable-atomic/issues/10#issuecomment-1159368067), and RISC-V and MIPS don't have double-width atomics, so those platforms are fine. I don't know about 32-bit powerpc/sparc's 64-bit atomic load -- IIRC LLVM does not support the former, and the latter is a tier 3 target.
[^ptr]: However, on 64-bit architecture's ILP32 ABIs, such as x86_64's [X32 ABI](https://en.wikipedia.org/wiki/X32_ABI), 64-bit load is definitely okay, so if Miri or rustc is aware of the architecture's pointer width, we may want to allow those as well.

Reply:
> Considering the above, I think it is ok to allow atomic load on read-only memory when all of the following are met. (assuming we don't want to do complex and fragile target-specific processing)

That's hoping we don't find further exceptions to these rules later? :D

We can possibly relax this rule later if there is consensus that all targets supported by Rust will support relaxed atomic loads of pointer size and smaller on read-only memory, but how sure are we no targets will violate this in the future?

Note that Miri anyway does not support sharing memory with other processes, so read-only memory here refers to `static` without interior mutability. Due to https://github.com/rust-lang/unsafe-code-guidelines/issues/303, this PR does not change behavior on any program that Miri currently accepts.

Reply:
Also let's discuss the Miri implementation here; the spec is being discussed at https://github.com/rust-lang/unsafe-code-guidelines/issues/355.